noun/trace.c

```diff
@@ -295,6 +298,180 @@ u3t_flee(void)
+          "\"args\": {\"name\": \"Event Processing\"}},\n",
+          nock_pid);
+  fprintf(trace_file,
+          "{\"name\": \"thread_sort_index\", \"ph\": \"M\", \"pid\": %d, "
+          "\"tid\": 1, \"args\": {\"sort_index\": 1}},\n",
+          nock_pid);
+  fprintf(trace_file,
+          "{\"name\": \"thread_name\", \"ph\": \"M\", \"pid\": %d, \"tid\": 2, "
+          "\"args\": {\"name\": \"Nock\"}},\n",
+          nock_pid);
+  fprintf(trace_file,
+          "{\"name\": \"thread_sort_index\", \"ph\": \"M\", \"pid\": %d, "
+          "\"tid\": 2, \"args\": {\"sort_index\": 2}},\n",
+          nock_pid);
+}
+
+/*  u3t_trace_time(): microsecond clock
+*/
+uint64_t u3t_trace_time()
+{
+  struct timeval tim_tv;
+  gettimeofday(&tim_tv, 0);
+  return 1000000ULL * tim_tv.tv_sec + tim_tv.tv_usec;
+}
+
+/* u3t_nock_trace_push(): push a trace onto the trace stack; returns yes if pushed.
+ *
+ * The trace stack is a stack of [path time-entered].
+ */
+c3_o
+u3t_nock_trace_push(u3_noun lab)
+{
+  if (!trace_file)
+    return c3n;
+
+  if ( (u3_nul == u3R->pro.trace) ||
+       !_(u3r_sing(lab, u3h(u3h(u3R->pro.trace)))) ) {
+    u3a_gain(lab);
+    c3_d time = u3t_trace_time();
+    u3R->pro.trace = u3nc(u3nc(lab, u3i_chubs(1, &time)), u3R->pro.trace);
+    return c3y;
+  }
+  else {
+    return c3n;
+  }
+}
+
+/*  _in_trace_pretty: measure/cut prettyprint.
+ *
+ *  Modeled after _cm_in_pretty(), the backend to u3m_p(), but with the
+ *  assumption that we're always displaying a path.
+ */
+static c3_w
+_in_trace_pretty(u3_noun som, c3_c* str_c)
+{
+  if ( _(u3du(som)) ) {
+    c3_w sel_w, one_w, two_w;
+    if ( str_c ) {
+      *(str_c++) = '/';
+    }
+    sel_w = 1;
+
+    one_w = _in_trace_pretty(u3h(som), str_c);
+    if ( str_c ) {
+      str_c += one_w;
+    }
+
+    two_w = _in_trace_pretty(u3t(som), str_c);
+    if ( str_c ) {
+      str_c += two_w;
+    }
+
+    return sel_w + one_w + two_w;
+  }
+  else {
+    c3_w len_w = u3r_met(3, som);
+    if ( str_c && len_w ) {
+      u3r_bytes(0, len_w, (c3_y *)str_c, som);
+      str_c += len_w;
+    }
+    return len_w;
+  }
+}
+
+static c3_c*
+trace_pretty(u3_noun som)
+{
+  c3_w len_w = _in_trace_pretty(som, NULL);
+  c3_c* pre_c = malloc(len_w + 1);
+
+  _in_trace_pretty(som, pre_c);
+  pre_c[len_w] = 0;
+  return pre_c;
+}
+
+/* u3t_nock_trace_pop(): pops a trace from the trace stack.
+ *
+ * When we remove the trace from the stack, we check to see if the sample is
+ * large enough to process, as we'll otherwise keep track of individual +add
+ * calls. If it is, we write it out to the tracefile.
+ */
+void
+u3t_nock_trace_pop()
+{
+  if (!trace_file)
+    return;
+
+  u3_noun trace  = u3R->pro.trace;
+  u3R->pro.trace = u3k(u3t(trace));
+
+  u3_noun item = u3h(trace);
+  u3_noun lab = u3h(item);
+  c3_d start_time = u3r_chub(0, u3t(item));
+
+  // 33microseconds (a 30th of a millisecond).
+  uint64_t duration = u3t_trace_time() - start_time;
+  if (duration > 33) {
```

Comment:
At some point, it would be nice to make this magic number configurable. Out of scope for this PR, though.

Reply:
ACK.